unrouted_handler.go

```diff
@@ -0,0 +1,626 @@
+type Config struct {
+	// DataStore implementation used to store and retrieve the single uploads.
+	// Must no be nil.
+	DataStore DataStore
+	// MaxSize defines how many bytes may be stored in one single upload. If its
+	// value is is 0 or smaller no limit will be enforced.
+	MaxSize int64
+	// BasePath defines the URL path used for handling uploads, e.g. "/files/".
+	// If no trailing slash is presented it will be added. You may specify an
+	// absolute URL containing a scheme, e.g. "http://tus.io"
+	BasePath string
+	// Initiate the CompleteUploads channel in the Handler struct in order to
+	// be notified about complete uploads
+	NotifyCompleteUploads bool
+	// Logger the logger to use internally
+	Logger *log.Logger
+}
+
+// UnroutedHandler exposes methods to handle requests as part of the tus protocol.
+// These are PostFile and PatchFile. It also includes GetFile and DelFile
+// however these are part of the spec. They are provided for convenience.
+type UnroutedHandler struct {
+	config        Config
+	dataStore     DataStore
+	isBasePathAbs bool
+	basePath      string
+	locks         map[string]bool
+	logger        *log.Logger
+
+	// For each finished upload the corresponding info object will be sent using
+	// this unbuffered channel. The NotifyCompleteUploads property in the Config
+	// struct must be set to true in order to work.
+	CompleteUploads chan FileInfo
+}
+
+// NewUnroutedHandler creates a new handler without routing using the given
+// configuration. It exposes the http handlers which need to be combined with
+// a router (aka mux) of your choice. If you are looking for preconfigured
+// handler see NewHandler.
+func NewUnroutedHandler(config Config) (*UnroutedHandler, error) {
+	logger := config.Logger
+	if logger == nil {
+		logger = log.New(os.Stdout, "[tusd] ", 0)
+	}
+	base := config.BasePath
+	uri, err := url.Parse(base)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure base path ends with slash to remove logic from absFileURL
+	if base != "" && string(base[len(base)-1]) != "/" {
+		base += "/"
+	}
+
+	// Ensure base path begins with slash if not absolute (starts with scheme)
+	if !uri.IsAbs() && len(base) > 0 && string(base[0]) != "/" {
+		base = "/" + base
+	}
+
+	handler := &UnroutedHandler{
+		config:          config,
+		dataStore:       config.DataStore,
+		basePath:        base,
+		isBasePathAbs:   uri.IsAbs(),
+		locks:           make(map[string]bool),
+		CompleteUploads: make(chan FileInfo),
+		logger:          logger,
+	}
+
+	return handler, nil
+}
+
+// TusMiddleware checks various aspects of the request and ensures that it
+// conforms with the spec. Also handles method overriding for clients which
+// cannot make PATCH AND DELETE requests. If you are using the tusd handlers
+// directly you will need to wrap at least the POST and PATCH endpoints in
+// this middleware.
+func (handler *UnroutedHandler) TusMiddleware(h http.Handler) http.Handler {
```

Comment:
I appreciate your example a lot since it helped me fundamentally to understand the purpose if your changes. What do you think or naming this method just `Middleware`?